pkg/auth/providers/genericoidc/genericoidc_provider.go

```diff
@@ -0,0 +1,165 @@
+package genericoidc
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	baseoidc "github.com/rancher/rancher/pkg/auth/providers/oidc"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	client "github.com/rancher/rancher/pkg/client/generated/management/v3"
+	publicclient "github.com/rancher/rancher/pkg/client/generated/management/v3public"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/rancher/pkg/user"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type GenOIDCProvider struct {
+	baseoidc.OpenIDCProvider
+}
+
+const (
+	Name      = "genericoidc"
+	UserType  = "user"
+	GroupType = "group"
+)
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager, tokenMGR *tokens.Manager) common.AuthProvider {
+	return &GenOIDCProvider{
+		baseoidc.OpenIDCProvider{
+			Name:        Name,
+			Type:        client.GenericOIDCConfigType,
+			CTX:         ctx,
+			AuthConfigs: mgmtCtx.Management.AuthConfigs(""),
+			Secrets:     mgmtCtx.Core.Secrets(""),
+			UserMGR:     userMGR,
+			TokenMGR:    tokenMGR,
+		},
+	}
+}
+
+// GetName returns the name of this provider.
+func (g *GenOIDCProvider) GetName() string {
+	return Name
+}
+
+// SearchPrincipals will return a principal of the requested principalType with a displayName and loginName
+// that match the searchValue.  This is done because OIDC does not have a proper lookup mechanism.  In order
+// to provide some degree of functionality that allows manual entry for users/groups, this is the compromise.
+func (g *GenOIDCProvider) SearchPrincipals(searchValue, principalType string, _ v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+
+	if principalType == "" {
+		principalType = UserType
+	}
+
+	p := v3.Principal{
+		ObjectMeta:    metav1.ObjectMeta{Name: g.Name + "_" + principalType + "://" + searchValue},
+		DisplayName:   searchValue,
+		LoginName:     searchValue,
+		PrincipalType: principalType,
+		Provider:      g.Name,
+	}
+
+	principals = append(principals, p)
+	return principals, nil
+}
+
+func (g *GenOIDCProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	var p v3.Principal
+
+	// parsing id to get the external id and type. Example genericoidc_<user|group>://<user sub | group name>
+	var externalID string
+	parts := strings.SplitN(principalID, ":", 2)
+	if len(parts) != 2 {
```

Comment:
Is it possible that the external id may contain `:`? If it's a possibility we can do something along the lines of.
```go
	principalScheme, externalID, found := strings.Cut(principalID, "://")
	...
	principalProvider, principalType, found := strings.Cut(principalScheme, "_")
	...
```

Reply:
Interesting question.  I've never seen a colon appear, but if we're aiming for "generic", we may want to be as safe as possible...just in case.